chia/full_node/weight_proof_v2.py

```diff
@@ -0,0 +1,1837 @@
+    curr_height = tip_height
+    blocks_n = 0
+    while ses_count < 2:
+        if curr_height == 0:
+            break
+        # add to needed reward chain recent blocks
+        header_block = headers[blockchain.height_to_hash(curr_height)]
+        block_rec = blocks[header_block.header_hash]
+        if header_block is None:
+            log.error("creating recent chain failed")
+            return None
+        recent_chain.insert(0, header_block)
+        if block_rec.sub_epoch_summary_included:
+            ses_count += 1
+        curr_height = uint32(curr_height - 1)
+        blocks_n += 1
+
+    header_block = headers[blockchain.height_to_hash(curr_height)]
+    recent_chain.insert(0, header_block)
+
+    log.info(
+        f"recent chain, "
+        f"start: {recent_chain[0].reward_chain_block.height} "
+        f"end:  {recent_chain[-1].reward_chain_block.height} "
+    )
+    return recent_chain
+
+
+def blue_boxed_end_of_slot(sub_slot: EndOfSubSlotBundle) -> bool:
+    if sub_slot.proofs.challenge_chain_slot_proof.normalized_to_identity:
+        if sub_slot.proofs.infused_challenge_chain_slot_proof is not None:
+            if sub_slot.proofs.infused_challenge_chain_slot_proof.normalized_to_identity:
+                return True
+        else:
+            return True
+    return False
+
+
+def _sample_sub_epoch(
+    start_of_epoch_weight: uint128,
+    end_of_epoch_weight: uint128,
+    weight_to_check: Optional[List[uint128]],
+) -> bool:
+    """
+    weight_to_check: List[uint128] is expected to be sorted
+    """
+    if weight_to_check is None:
+        return True
+    if weight_to_check[-1] < start_of_epoch_weight:
+        return False
+    if weight_to_check[0] > end_of_epoch_weight:
+        return False
+    choose = False
+    for weight in weight_to_check:
+        if weight > end_of_epoch_weight:
+            return False
+        if start_of_epoch_weight < weight < end_of_epoch_weight:
+            log.debug(f"start weight: {start_of_epoch_weight}")
+            log.debug(f"weight to check {weight}")
+            log.debug(f"end weight: {end_of_epoch_weight}")
+            choose = True
+            break
+
+    return choose
+
+
+def _validate_recent_blocks(
+    constants_dict: Dict[str, Any], recent_chain_bytes: bytes, summaries_bytes: List[bytes]
+) -> Tuple[bool, List[bytes]]:
+    """
+    validate pospace for blocks after the first two slots, full validation for last 100 blocks
+    returns the result and the list of blocks records as bytes
+    """
+    constants, summaries = bytes_to_vars(constants_dict, summaries_bytes)
+    recent_chain: RecentChainData = RecentChainData.from_bytes(recent_chain_bytes)
+    sub_blocks = BlockCache({})
+    first_ses_idx = _get_ses_idx(recent_chain.recent_chain_data)
+    ses_idx = len(summaries) - len(first_ses_idx)
+    ssi: uint64 = constants.SUB_SLOT_ITERS_STARTING
+    diff: Optional[uint64] = constants.DIFFICULTY_STARTING
+    last_blocks_to_validate = 100
```

Comment:
this takes lots of cpu and runs in parallel to the sub epoch validation, increasing it will make the validation times longer, ill move this to the handler params for better visibilit.

Reply:
some numbers, each extra 100 blocks we validate adds about 10 sec to the validation